plugins/protocols/xdg-activation.cpp

```diff
@@ -12,53 +12,135 @@
+        last_token = nullptr;
+    }
 
     bool is_unloadable() override
     {
         return false;
     }
 
   private:
-    static void xdg_activation_handle_request_activate(struct wl_listener *listener, void *data)
+    void set_callbacks()
     {
-        auto event = static_cast<const struct wlr_xdg_activation_v1_request_activate_event*>(data);
-
-        wayfire_view view = wf::wl_surface_to_wayfire_view(event->surface->resource);
-        if (!view)
+        xdg_activation_request_activate.set_callback([this] (void *data)
         {
-            LOGE("Could not get view");
-            return;
-        }
+            auto event = static_cast<const struct wlr_xdg_activation_v1_request_activate_event*>(data);
 
-        auto toplevel = wf::toplevel_cast(view);
-        if (!toplevel)
-        {
-            LOGE("Could not get toplevel view");
-            return;
-        }
+            if (!event->token->seat)
+            {
+                LOGI("Denying focus request, token was rejected at creation");
+                return;
+            }
+
+            if (only_last_token && (event->token != last_token))
+            {
+                LOGI("Denying focus request, token is expired");
+                return;
+            }
+
+            last_token = nullptr; // avoid reusing the same token
 
-        if (!event->token->seat)
+            wayfire_view view = wf::wl_surface_to_wayfire_view(event->surface->resource);
+            if (!view)
+            {
+                LOGE("Could not get view");
+                return;
+            }
+
+            auto toplevel = wf::toplevel_cast(view);
+            if (!toplevel)
+            {
+                LOGE("Could not get toplevel view");
+                return;
+            }
+
+            LOGI("Activating view");
+            wf::get_core().default_wm->focus_request(toplevel);
+        });
+
+        xdg_activation_new_token.set_callback([this] (void *data)
         {
-            LOGI("Denying focus request, seat wasn't supplied");
-            return;
-        }
+            auto token = static_cast<struct wlr_xdg_activation_token_v1*>(data);
+            bool reject_token = false;
+            if (!token->seat)
+            {
+                // note: for a valid seat, wlroots already checks that the serial is valid
+                LOGI("Not registering activation token, seat was not supplied");
+                reject_token = true;
+            }
+
+            if (check_surface && !token->surface)
+            {
+                // note: for a valid surface, wlroots already checks that this is the active surface
+                LOGI("Not registering activation token, surface was not supplied");
+                token->seat  = nullptr; // this will ensure that this token will be rejected later
+                reject_token = true;
+            }
+
+            if (reject_token)
+            {
+                if (token == last_token)
```

Comment:
While this technically works, why not simply use the token's `destroy` event? This here (+leaving the pointer dangling) seems like quite the hack.

Reply:
I think I had a reason for it, but cannot remember what it was :)
Although I originally did something more complicated and might be a leftover from that. I'll try again and see.

Reply:
OK, I tried and it seems to work well with the destroy signal as well